source/common/network/dns_impl.cc

```diff
@@ -242,5 +255,90 @@ void DnsResolverImpl::PendingResolution::getAddrInfo(int family) {
       this);
 }
 
+ActiveDnsQuery* DnsResolverImpl::resolveSrv(const std::string& dns_name,
+                                            DnsLookupFamily dns_lookup_family, ResolveCb callback) {
+  std::unique_ptr<PendingSrvResolution> pending_srv_res(
+      new PendingSrvResolution(callback, dispatcher_, channel_, dns_name, dns_lookup_family, this));
+  pending_srv_res->getSrvByName();
+  return preparePendingResolution(std::move(pending_srv_res));
+}
+
+void DnsResolverImpl::PendingSrvResolution::onAresSrvStartCallback(int status, int timeouts,
+                                                                   unsigned char* buf, int len) {
+  // We receive ARES_EDESTRUCTION when destructing with pending queries.
+  if (status == ARES_EDESTRUCTION) {
+    ASSERT(owned_);
+    delete this;
+    return;
+  }
+
+  bool replies_parsed = false;
+  if (status == ARES_SUCCESS) {
+    struct ares_srv_reply* srv_reply;
+    status = ares_parse_srv_reply(buf, len, &srv_reply);
+
+    if (status == ARES_SUCCESS) {
+      size_t total = 0;
+      for (ares_srv_reply* current_reply = srv_reply; current_reply != NULL;
+           current_reply = current_reply->next) {
+        total++;
+      }
+
+      std::shared_ptr<std::atomic<ssize_t>> finished = std::make_shared<std::atomic<ssize_t>>(0);
+      std::shared_ptr<std::list<DnsResponse>> srv_records =
+          std::make_shared<std::list<DnsResponse>>();
+      std::shared_ptr<std::mutex> mutex = std::make_shared<std::mutex>();
+      for (ares_srv_reply* current_reply = srv_reply; current_reply != NULL;
+           current_reply = current_reply->next) {
+        resolver_->resolve(
+            current_reply->host, this->dns_lookup_family_,
+            [this, total, finished, srv_records, mutex,
+             current_reply](const std::list<DnsResponse>&& response) {
+              for (auto instance = response.begin(); instance != response.end(); ++instance) {
+                Address::InstanceConstSharedPtr inst_with_port(
+                    Utility::getAddressWithPort(*instance->address_, current_reply->port));
+                mutex->lock();
```

Comment:
Why are we taking locks? Isn't everything thread local on this dispatcher?

Reply:
Are threads and dispatchers always coupled?  I think so, but I'm not 100% sure.  So any operations happening through a particular dispatcher will operate on the same thread?

Reply:
For the purpose of DNS resolution, this should be true (that they are coupled).